[TableGen] Add support for emitting new function definition to return a range of results for Primary Key

quic-garvgupt opened the pull request:
In the RISC-V architecture, multiple vendor-specific Control and Status Registers (CSRs) share the same encoding. However, the existing lookup function, which currently returns only a single result, falls short. During disassembly, it consistently returns the first CSR encountered, which may not be the correct CSR for the subtarget.

To address this issue, we modify the function definition to return a range of results. These results can then be iterated upon to identify the CSR that best fits the subtarget’s feature requirements. The behavior of this new definition is controlled by a variable named `ReturnRange`, which defaults to `false`.

Specifically, this patch introduces support for emitting a new lookup function for the primary key. This function returns a pair of iterators pointing to the first and last values, providing a comprehensive range of values that satisfy the query

quic-garvgupt commented:
Hi, @wangpc-pp , @apazos, @asb, @topperc , can you please review this PR? Thanks

Re: for creating separate PR for changes in searchable Tables @wangpc-pp 
>> I understand that the changes will affect all targets, however before sending this PR to more broader set of reviewers, I wanted to know the comments from the reviewers of RISCV community because the framework here is to resolve encoding conflicts among CSRs, which as per my understanding in unique to RISCV. If and when this change is approved for RISCV, I will push it to a different PR with a broad set of reviewers.

topperc commented on llvm/utils/TableGen/SearchableTableEmitter.cpp:
```diff
@@ -452,55 +461,80 @@ void SearchableTableEmitter::emitLookupFunction(const GenericTable &Table,
+          "Key,\n";
+    OS << "    [](const KeyType &LHS, const " << IndexTypeName << " &RHS) {\n";
+    emitComparator();
+    OS << "  while(Idx != UpperBound){\n";
```
Need space after `while` and before `{`

topperc commented:
Can `lookupSysRegByEncoding` use `equal_range` instead of calling `lower_bound` and `upper_bound`? Then iterate the returned range to push into the vector?

topperc commented:
Can the new signature of `lookupSysRegByEncoding` return the std::pair from `equal_range` instead of needing to copy into a vector?

topperc commented on llvm/utils/TableGen/SearchableTableEmitter.cpp:
```diff
@@ -540,10 +577,20 @@ void SearchableTableEmitter::emitGenericTable(const GenericTable &Table,
+    // lookup signature for the primary key lookup : first will return a SysReg
+    // that matches the primary key, second will populate a vector passed as
+    // argument with all the SysRegs that match the primary key.
+    emitLookupFunction(Table, Index, /*IsPrimary=*/true,
```
Rather than emitting two different functions, can we have a flag in tablegen that indicates whether a table should have a single result or a range of results.

apazos commented:
In past community discussion we talked about how vendor CSRs will require an extension guarding them to be able to resolve conflicting encoding. Hence in this patch only the framework for handling conflicting encoding was pushed.

@topperc, thanks for the suggestions. 

It should be possible to return an iterator range instead of passing in a vector to be populated.

The decision to emit 2 APIs was to not break targets that already use the original API. This allows making a toolchain build that supports more than one target. I also understand only the PrimaryKey lookup function does the lookup by encoding. It is possible to add a variable to GenericTable to control the return value type of the PrimaryKey lookup function, but not sure it helps much. We build several targets, some with custom CSRs (like in RISC-V downstream toolchains) and targets that do not have them (e..g, ARM, AArch64, etc.) and need the original API. So at the end, the tablegen generated code will have to produce both APIs.

@quic-garvgupt, it will be great if you can add a tablegen test showing the usage example you have in the commit message.

topperc commented:
> We build several targets, some with custom CSRs (like in RISC-V downstream toolchains) and targets that do not have them (e..g, ARM, AArch64, etc.) and need the original API. So at the end, the tablegen generated code will have to produce both APIs.

Does the tablegen code being modified know about CSRs? Isn't it just a generic framework for a searchable table?

My suggestion was that RISCV would set the flag in its .td file to make the CSR table use a range. No other target needs to be affected since they won't set the flag.

jrtc27 commented:
> > We build several targets, some with custom CSRs (like in RISC-V downstream toolchains) and targets that do not have them (e..g, ARM, AArch64, etc.) and need the original API. So at the end, the tablegen generated code will have to produce both APIs.
> 
> Does the tablegen code being modified know about CSRs? Isn't it just a generic framework for a searchable table?
> 
> My suggestion was that RISCV would set the flag in its .td file to make the CSR table use a range. No other target needs to be affected since they won't set the flag.

Per-table though, please, i.e. a bit in SearchIndex and a corresponding PrimaryKeyFoo one in GenericTable.

jrtc27 commented on llvm/utils/TableGen/SearchableTableEmitter.cpp:
```diff
@@ -540,10 +577,20 @@ void SearchableTableEmitter::emitGenericTable(const GenericTable &Table,
+  if (Table.PrimaryKey) {
+    auto &Index = *Table.PrimaryKey;
+    // Two lookupfunction functions need to be generated to allow more than one
+    // lookup signature for the primary key lookup : first will return a SysReg
```
SysReg is a specific use in the RISC-V target of this generic feature

topperc commented:
> > > We build several targets, some with custom CSRs (like in RISC-V downstream toolchains) and targets that do not have them (e..g, ARM, AArch64, etc.) and need the original API. So at the end, the tablegen generated code will have to produce both APIs.
> > 
> > 
> > Does the tablegen code being modified know about CSRs? Isn't it just a generic framework for a searchable table?
> > My suggestion was that RISCV would set the flag in its .td file to make the CSR table use a range. No other target needs to be affected since they won't set the flag.
> 
> Per-table index though, please, i.e. a bit in SearchIndex and a corresponding PrimaryKeyFoo one in GenericTable.

Agreed. It should be like `PrimaryKeyEarlyOut` in `GenericTable` and `EarlyOut` in `SearchIndex`.

apazos commented:
It is fine to add the variables to GenericTable and SearchIndex to control the API being generated. 

When upstreaming the framework the values of those variables will be set to false by default, since upstream RISC-V does not have vendor CSRs that can conflict. 

Downstream users who have custom/vendor CSRs will set them to true but will require additional changes to call the proper API in RISCVAsmParser::parseCSRSystemRegister and RISCVInstPrinter::printCSRSystemRegister. 

Maybe we can consider a separate patch that sets the variables to true for RISC-V and modifies parser/printer to use the new API. This way downstream users of RISC-V backend do not have the burden to maintain those changes. I think most downstream users have vendor/custom CSRs that have encoding conflict. Would that be acceptable?

topperc commented:
> Maybe we can consider a separate patch that sets the variables to true for RISC-V and modifies parser/printer to use the new API. This way downstream users of RISC-V backend do not have the burden to maintain those changes. I think most downstream users have vendor/custom CSRs that have encoding conflict. Would that be acceptable?

Yes. I definitely think we should do that.

Commit:
[RISCV]Add support for resolving encoding conflicts among vendor specific CSRs

This patch adds the framework for resolving encoding conflicts among CSRs.

Specifically, this patch adds a support for emitting a new lookup function for
the primary key which return a pair of iterators pointing to first and last
value hence giving a range of values which satisfies the query.

While printing the CSR name during objdump, iterate over the range and print the
name of only that CSR which satisifes the feature requirement of subtarget.

Below is the signature of the new function that will be emitted for primary key:

```
llvm::iterator_range<const SysReg *> lookupSysRegByEncoding(uint16_t Encoding) {
  struct KeyType {
    uint16_t Encoding;
  };
  KeyType Key = {Encoding};
  struct Comp {
    bool operator()(const SysReg &LHS, const KeyType &RHS) const {
      if (LHS.Encoding < RHS.Encoding)
        return true;
      if (LHS.Encoding > RHS.Encoding)
        return false;
      return false;
    }
    bool operator()(const KeyType &LHS, const SysReg &RHS) const {
      if (LHS.Encoding < RHS.Encoding)
        return true;
      if (LHS.Encoding > RHS.Encoding)
        return false;
      return false;
    }
  };
  auto Table = ArrayRef(SysRegsList);
  auto It = std::equal_range(Table.begin(), Table.end(), Key, Comp());
  return llvm::make_range(It.first, It.second);
}
```

NOTE: Emitting a different signature for returning a range of results is only
supported by primary key.

topperc commented on llvm/test/TableGen/EncodingConflict.td:
```diff
@@ -0,0 +1,118 @@
+// CHECK-NEXT:     });
+
+// CHECK:       auto Idx = It.first;
+// CHECK-NEXT:  if (Idx == Table.end() ||
```
Is this needed? If there are no values in the table that contain the value we're looking for, then both iterators will end up the same, making an empty range.

quic-garvgupt replied:
I have made the changes in the latest patchset.

topperc commented on llvm/test/TableGen/EncodingConflict.td:
```diff
@@ -0,0 +1,118 @@
+// CHECK-NEXT:   { "csr2", 0x7C9,  {Feature2}  }, // 1
+// CHECK-NEXT:  };
+
+// CHECK: std::pair<const SysReg *, const SysReg *>lookupSysRegByEncoding(uint16_t Encoding) {
```
Would it be better to return `llvm::iterator_range` instead of `std::pair`? I think that would make it naturally compatible with range based loops in the caller.

jrtc27 commented on llvm/lib/Target/RISCV/RISCVSystemOperands.td:
```diff
@@ -49,6 +49,7 @@ def SysRegsList : GenericTable {
 
   let PrimaryKey = [ "Encoding" ];
   let PrimaryKeyName = "lookupSysRegByEncoding";
+  let PrimaryKeyReturnRange = false;
```
This is pointless, just add the line when you set it to true

jrtc27 commented on llvm/include/llvm/TableGen/SearchableTable.td:
```diff
@@ -135,6 +138,11 @@ class SearchIndex {
   bit EarlyOut = false;
+
+  // If true, will generate a different function signature which will return
+  // a std::pair<> of iterators pointing to start and end value of the range
```
Not true any more

jrtc27 commented on llvm/test/TableGen/EncodingConflict.td:
```diff
@@ -0,0 +1,109 @@
+// CHECK: #undef GET_List1_DECL
+// CHECK-NEXT: #undef GET_List1_IMPL
+// CHECK-NEXT: #undef GET_List2_DECL
+// CHECK-NEXT: #undef GET_List2_IMPL
```
Fix

jrtc27 commented on llvm/test/TableGen/EncodingConflict.td:
```diff
@@ -0,0 +1,109 @@
+  let PrimaryKeyName = "lookupSysRegByEncoding";
+  let PrimaryKeyReturnRange = true;
+}
+
+
```
```suggestion

```

jrtc27 commented on llvm/utils/TableGen/SearchableTableEmitter.cpp:
```diff
@@ -478,25 +500,34 @@ void SearchableTableEmitter::emitLookupFunction(const GenericTable &Table,
-  OS << "const " << Table.CppTypeName << " *" << Index.Name << "(";
-
+  if (ShouldReturnRange)
+    OS << "llvm::iterator_range<const " << Table.CppTypeName << " *>";
```
```suggestion
    OS << "llvm::iterator_range<const " << Table.CppTypeName << " *> ";
```

jrtc27 commented on llvm/utils/TableGen/SearchableTableEmitter.cpp:
```diff
@@ -190,15 +191,18 @@ class SearchableTableEmitter {
   void emitGenericEnum(const GenericEnum &Enum, raw_ostream &OS);
   void emitLookupDeclaration(const GenericTable &Table,
-                             const SearchIndex &Index, raw_ostream &OS);
+                             const SearchIndex &Index, bool ShouldReturnRange,
```
At this point I believe all the callers of this and emitLookupFunction pass Index.ReturnRange for the parameter, so it's unneeded?

topperc commented on llvm/include/llvm/TableGen/SearchableTable.td:
```diff
@@ -135,6 +138,11 @@ class SearchIndex {
   // Can only be used when the first field is an integral (non-string) type.
   bit EarlyOut = false;
+
+  // If true, will generate a different function signature which will return a
```
"a iterator range" -> "an iterator range"

topperc commented on llvm/utils/TableGen/SearchableTableEmitter.cpp:
```diff
@@ -426,16 +428,25 @@ void SearchableTableEmitter::emitLookupFunction(const GenericTable &Table,
-    OS << "    " << searchableFieldType(Table, Index, Field, TypeInTempStruct)
-       << " " << Field.Name << ";\n";
+  bool ShouldReturnRange = Index.ReturnRange;
+  if (ShouldReturnRange)
```
Why does returning a range need to change how the Key is constructed?

quic-garvgupt replied:
std::equal_range expects both arguments to be of same type unlike in std::lower_bound where both can be of different type.

auto Idx = std::lower_bound(Table.begin(), Table.end(), Key,
    [](const **SysReg** &LHS, const **KeyType** &RHS) {
    
auto It = std::equal_range(Table.begin(), Table.end(), Key,
    [](const **SysReg** &LHS, const **SysReg** &RHS) {

topperc replied:
Here's a patch to fix that in a better way. https://github.com/llvm/llvm-project/pull/97784

quic-garvgupt replied:
Thanks! Should I include those changes in my next patchset or will it be merged separately?

topperc replied:
You can just include them in your next patchset. Only emit the second compare function I added if `ReturnRange` is true.

quic-garvgupt replied:
I understand that this will simplify the code changes for generating the key, however we would still need to use equal_range instead of lower_bound to return a range of answers like below. Do let me know if this is what you were expecting
```
llvm::iterator_range<const SysReg *> lookupSysRegByEncoding(uint16_t Encoding) {
  struct KeyType {
    uint16_t Encoding;
  };
  KeyType Key = {Encoding};
  struct Comp {
    bool operator()(const SysReg &LHS, const KeyType &RHS) const {
      if (LHS.Encoding < RHS.Encoding)
        return true;
      if (LHS.Encoding > RHS.Encoding)
        return false;
      return false;
    }
    bool operator()(const KeyType &LHS, const SysReg &RHS) const {
      if (LHS.Encoding < RHS.Encoding)
        return true;
      if (LHS.Encoding > RHS.Encoding)
        return false;
      return false;
    }
  };
  auto Table = ArrayRef(SysRegsList);
  auto It = std::equal_range(Table.begin(), Table.end(), Key, Comp());
  return llvm::make_range(It.first, It.second);
}
```

topperc replied:
Yes that is what I was expecting.

topperc commented on llvm/utils/TableGen/SearchableTableEmitter.cpp:
```diff
@@ -510,7 +538,8 @@ void SearchableTableEmitter::emitGenericTable(const GenericTable &Table,
   // Emit the declarations for the functions that will perform lookup.
   if (Table.PrimaryKey) {
-    emitLookupDeclaration(Table, *Table.PrimaryKey, OS);
+    auto &Index = Table.PrimaryKey;
```
Why is this change needed?

quic-garvgupt replied:
This was needed before when an additional argument `ShouldReturnRange` was being passed to emitLookupFunction. But since, it was not needed anymore, we can remove this extra assignment statement as well. I have pushed a new patchset fixing the same

topperc commented on llvm/utils/TableGen/SearchableTableEmitter.cpp:
```diff
@@ -540,10 +569,12 @@ void SearchableTableEmitter::emitGenericTable(const GenericTable &Table,
-  if (Table.PrimaryKey)
-    emitLookupFunction(Table, *Table.PrimaryKey, true, OS);
+  if (Table.PrimaryKey) {
+    auto &Index = Table.PrimaryKey;
```
Why do we need this reference?

quic-garvgupt replied:
This was needed before when an additional argument `ShouldReturnRange` was being passed to emitLookupFunction. But since, it was not needed anymore, we can remove this extra assignment statement as well. I have pushed a new patchset fixing the same

topperc commented on llvm/test/TableGen/EncodingConflict.td:
```diff
@@ -0,0 +1,108 @@
+// RUN: llvm-tblgen -gen-searchable-tables -I %p/../../include %s | FileCheck %s
```
Can you rename this test to refer to the feature being tested? Not a particular usage of the feature.

quic-garvgupt replied:
Made it to ReturnRange.td

topperc commented on llvm/utils/TableGen/SearchableTableEmitter.cpp:
```diff
@@ -426,16 +428,25 @@ void SearchableTableEmitter::emitLookupFunction(const GenericTable &Table,
-    OS << "    " << searchableFieldType(Table, Index, Field, TypeInTempStruct)
-       << " " << Field.Name << ";\n";
+  bool ShouldReturnRange = Index.ReturnRange;
+  if (ShouldReturnRange)
```
Is `ShouldReturnRange` only compatible with `IsPrimary==true`?

quic-garvgupt replied:
IIUC your comment - `ShouldReturnRange` can be `true` irrespective of `IsPrimary` is true or not. 
`ShouldReturnRange` is populated by the `ReturnRange` field of `SearchIndex`. So if `ReturnRange` is set to true for a non-primary function, then also `ShouldReturnRange` will hold true. An example to illustrate the same is already added in `ReturnRange.td` file.

topperc commented on llvm/utils/TableGen/SearchableTableEmitter.cpp:
```diff
@@ -478,25 +498,33 @@ void SearchableTableEmitter::emitLookupFunction(const GenericTable &Table,
-  if (IsPrimary)
+  if (ShouldReturnRange)
+    OS << "  return llvm::make_range(It.first, It.second);\n";
+  else if (IsPrimary) {
```
@quic-garvgupt I asked the question about `IsPrimary` because of this code that has to index the table using `Idx->_index` when the `IsPrimary` is false.

topperc commented on llvm/test/TableGen/ReturnRange.td:
```diff
@@ -0,0 +1,108 @@
+// CHECK-NEXT:       return false;
+// CHECK-NEXT:     });
+
+// CHECK:   return llvm::make_range(It.first, It.second);
```
The iterator here is `IndexType *` right? That's not convertible to `llvm::iterator_range<const SysReg *>` is it?

quic-garvgupt replied:
Thanks for pointing this out!. I have fixed this in the latest patchset. This required emitting a separate global table sorted by the field which we are trying to lookup with (for example, in the test it is `Name`), so that a range of entries can be returned. Do let me know what you think of it. 

Though, there is currently no need of emitting a different signature for non-primary functions both upstream and in downstream I think, therefore if it is felt that there is so much of redundant code-changes, then at this point maybe we can just support emitting different signature for primary functions only.

topperc replied:
I think we can drop the support for ReturnRange and !PrimaryKey for now.

quic-garvgupt replied:
Just to be clear, I should only allow emitting different signature from primary lookupfunctions and not support emitting different signature for functions other than primary lookupfunctions (eg. `lookupsysregbyname`). 
For primary function , there is already a use case of encoding conflict, so would need to emit a different signature. However, there is no use case for emitting a different signature for non-primary functions(eg. `lookupsysregbyname`), so probably the added complexity is not worth it for the time being.

Apologies if I misunderstood your comment or missed something. This comment is just to clarify on what needs to  be dropped and what can be retained.

topperc replied:
`Just to be clear, I should only allow emitting different signature from primary lookupfunctions and not support emitting different signature for functions other than primary lookupfunctions`

Yes.

Sorry I meant to write `IsPrimary` instead of `PrimaryKey` in my reply. I fixed it.

topperc commented:
Please retitle this patch and update the description. The only thing in this patch is a new tablegen feature that should be described on its own. The RISC-V use case can be used to justify it, but should not be the primary description.

topperc commented on llvm/utils/TableGen/SearchableTableEmitter.cpp:
```diff
@@ -538,12 +573,41 @@ void SearchableTableEmitter::emitGenericTable(const GenericTable &Table,
+  // Emit Table for all those indices which will return a range instead of a
+  // single record
+  for (const auto &Index : Table.Indices) {
+    if (Index->ReturnRange) {
```
Invert the condition and use a `continue to reduce nesting

topperc commented on llvm/utils/TableGen/SearchableTableEmitter.cpp:
```diff
@@ -538,12 +573,41 @@ void SearchableTableEmitter::emitGenericTable(const GenericTable &Table,
+      std::vector<Record *> Entries;
+      Entries.reserve(Table.Entries.size());
+      for (unsigned i = 0; i < Table.Entries.size(); ++i)
+        Entries.emplace_back(Table.Entries[i]);
```
Does `std::vector<Record *> Entries(Table.Entries.begin(), Table.Entries.end()` work instead of writing your own copy loop?

nvjle commented:
Please also update the _SearchableTables Reference_  in `docs/TableGen/BackEnds.rst`.

topperc commented on llvm/utils/TableGen/SearchableTableEmitter.cpp:
```diff
@@ -848,7 +884,7 @@ void SearchableTableEmitter::run(raw_ostream &OS) {
           (Twine("lookup") + Table->CppTypeName + "By" + Field).str();
       Table->Indices.push_back(parseSearchIndex(*Table, Class->getValue(Field),
-                                                Name, {Field}, false));
+                                                Name, {Field}, false, false));
```
Label these two `false` values with an inline comment.

topperc commented on llvm/include/llvm/TableGen/SearchableTable.td:
```diff
@@ -135,6 +138,12 @@ class SearchIndex {
   //
   // Can only be used when the first field is an integral (non-string) type.
   bit EarlyOut = false;
+
```
Should we remove this since it always generate an error?

quic-garvgupt replied:
The primarykey field in genericTable is an instance of `SearchIndex` and is parsed using `parseSearchIndex` which is also used for parsing non-primary functions. Therefore it is necessary to keep `ReturnRange` in `SearchIndex` even though it will always throw an error.

quic-garvgupt commented:
I am not able to install the dependencies needed to build the documentation locally. I will really appreciate if someone else can build the documentation and verify if the changes in `BackEnds.rst` file are fine. Thanks

francisvm commented:
> I am not able to install the dependencies needed to build the documentation locally. I will really appreciate if someone else can build the documentation and verify if the changes in `BackEnds.rst` file are fine. Thanks

The "Test documentation build" check below does `ninja docs-llvm-html docs-llvm-man`, and looks successful, so I think you're good.

quic-garvgupt commented:
> > I am not able to install the dependencies needed to build the documentation locally. I will really appreciate if someone else can build the documentation and verify if the changes in `BackEnds.rst` file are fine. Thanks
> 
> The "Test documentation build" check below does `ninja docs-llvm-html docs-llvm-man`, and looks successful, so I think you're good.

I didn't realize we build documentation. Thanks for letting me know!

wangpc-pp commented on llvm/test/TableGen/ReturnRange.td:
```diff
@@ -0,0 +1,108 @@
+// RUN: llvm-tblgen -gen-searchable-tables -I %p/../../include %s | FileCheck %s
```
Maybe rename this file to `llvm/test/TableGen/generic-tables-return-range.td`.

wangpc-pp reviewed (approved):
LGTM. I think all issues have been fixed?

topperc commented on llvm/utils/TableGen/SearchableTableEmitter.cpp:
```diff
@@ -793,7 +828,8 @@ void SearchableTableEmitter::run(raw_ostream &OS) {
                          IndexRec->getValueAsListOfStrings("Key"),
-                         IndexRec->getValueAsBit("EarlyOut")));
+                         IndexRec->getValueAsBit("EarlyOut"),
+                         IndexRec->getValueAsBit("ReturnRange")));
```
Can we just pass `false` here instead of calling `IndexRec->getValueAsBit("ReturnRange")`? Would that remove the need for having `ReturnRange` in the `SearchIndex` class in tablegen?

quic-garvgupt replied:
This would still require to have `ReturnRange` field in `struct SearchIndex` defined in the SearchableTableEmitter file. I have made the changes. Do let me know if this will be fine

quic-garvgupt replied:
I do not have commit access yet, if changes are good, can you merge this PR and PR#97287 on my behalf?

topperc replied:
> This would still require to have `ReturnRange` field in `struct SearchIndex` defined in the SearchableTableEmitter file. I have made the changes. Do let me know if this will be fine

That's fine. I just didn't want to expose something in the .td file that would always error if we could avoid it. The class in SearchTableEmitter isn't as user facing as the .td file.

topperc reviewed (approved):
LGTM
